### PR TITLE
Defer: move the stream of responses inside ExecutionResponse

### LINKED ISSUE #1188: Defer: move the stream of responses inside ExecutionResponse and RouterResponse
Instead of using `BoxStream<ExecutionResponse>` everywhere, we could move the stream inside of `ExecutionResponse`, as follows:

```diff
-pub struct ExecutionResponse {                             
-    pub response: http_compat::Response<Response>,         
+pub struct ExecutionResponse<T: Stream<Item = Response>> { 
+    pub response: http_compat::Response<T>,                
                                                            
     pub context: Context,                                  
 }                                                                                                         
```

This is explored in the branch https://github.com/apollographql/router/tree/geal/les-epinards-cest-plein-defer

Right now this is running into multiple issues:
- https://github.com/BrynCooke/buildstructor/issues/60
- we cannot make `ExecutionResponse` generic over the stream in `DynPlugin::execution_service`, because then the trait is not object safe: https://github.com/apollographql/router/blob/geal/les-epinards-cest-plein-defer/apollo-router-core/src/plugin/mod.rs#L128-L133

### PR DESCRIPTION
**Pull request by Geal**
Fix #1188 

This makes the API nicer to use while also matching more precisely the behaviour of the HTTP server: headers are sent once at the beginning of the response instead of each deferred response

TODO:
- [x] same change for `RouterResponse`
- [ ] separate callbacks for headers and responses for execution and route rhai plugins
- [ ] evaluate a more generic API to avoid boxes. If we cannot avoid boxes, `ExecutionResponse` can just contain a `BoxStream` explicitely instead of being generic

### CODE REVIEW

**Comment by BrynCooke** [comment]
I feel this PR is moving things in the right direction. 

We should do avoid another release of the router until this work is completed. We will need to provide migration steps for existing users, and ideally we want to limit the number of times we make them make major changes.

File: apollo-router/src/services/mod.rs [hunk 1]
```diff
@@ -608,29 +639,18 @@ impl ExecutionRequest {
     }
 }
 
-assert_impl_all!(ExecutionResponse: Send);
+//assert_impl_all!(ExecutionResponse: Send);
```
Comment (BrynCooke) [inline apollo-router/src/services/mod.rs:642]: Probably remove rather than just commenting out.

**Review by BrynCooke — APPROVED** [review]
I think this is much better than where we were.

My only gripe is around the parameterization of RouterResponse to avoid boxing for when there is guaranteed to only be one response. It can be revisited after this is merged though, as I think in general this is getting us to a much better place.

**Comment by Geal** [comment]
> My only gripe is around the parameterization of RouterResponse to avoid boxing for when there is guaranteed to only be one response. It can be revisited after this is merged though, as I think in general this is getting us to a much better place.

The response types could be changed to something like:
```rust
struct RouterResponse<T: Stream<Item = ResponseBody>> {
    context: Context,
    response: ResponseBody,
    stream: Option<T>,
}
```

File: apollo-router/src/services/mod.rs (hunk 1 above)
Comment (Geal) [reply apollo-router/src/services/mod.rs]: added back in [440cdc0](https://github.com/apollographql/router/pull/1206/commits/440cdc00f5109c0843e70ad0ade050e4b07bde36)

**Comment by BrynCooke** [comment]
> > My only gripe is around the parameterization of RouterResponse to avoid boxing for when there is guaranteed to only be one response. It can be revisited after this is merged though, as I think in general this is getting us to a much better place.
> 
> The response types could be changed to something like:
> 
> ```rust
> struct RouterResponse<T: Stream<Item = ResponseBody>> {
>     context: Context,
>     response: ResponseBody,
>     stream: Option<T>,
> }
> ```

I think it is better to go with what you have right now and maybe revisit when we remove ResponseBody as part of the api-1.0 work. Once this is done we can use buildstructor in combination with an enum to allow users to construct either a stream or single response without any parameterization.

**Pull request merged**